src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts

```diff
@@ -327,19 +350,59 @@ class RemoteMachineTrainingService implements TrainingService {
         return deferred.promise;
     }
 
-    public cleanUp(): Promise<void> {
+    public async cleanUp(): Promise<void> {
         this.log.info('Stopping remote machine training service...');
         this.stopping = true;
+        let timeoutId: NodeJS.Timer;
+        const delay: Promise<{}> = new Promise((resolve: Function, reject: Function): void => {
+            timeoutId = setTimeout(
+                () => { reject(new Error('TrainingService cleanup timeout.')); },
+                10000);
+        });
+        await Promise.race([delay, this.cleanupConnections()]).finally(() => {
+            clearTimeout(timeoutId);
+        });
+    }
+    
+    /**
+     * stop gpu_metric_collector process in remote machine and remove unused scripts
+     */
+    private async cleanupConnections(): Promise<void> {
+        try{
+            for (const [rmMeta, client] of this.machineSSHClientMap.entries()) {
+                let jobpidPath: string = path.join(this.getRemoteScriptsPath(), 'pid');
+                await SSHClientUtility.remoteExeCommand(`pkill -P \`cat ${jobpidPath}\``, client);
+                await SSHClientUtility.remoteExeCommand(`rm -rf ${this.getRemoteScriptsPath()}`, client);
+            }
+        }catch (error) {
+            this.log.error(`Cleanup connection exception, error is ${error.message}`);
+        }
 
         return Promise.resolve();
+    } 
+    
+    private async generateGpuScript(): Promise<string> {
+        let localGpuMetricFolder: string = `/tmp/nni/scripts/${uniqueString(5)}`;
+        await cpp.exec(`mkdir -p ${localGpuMetricFolder}`);
+        //generate gpu_metrics_collector.sh
+        let gpuMetricFilePath: string = path.join(localGpuMetricFolder, 'gpu_metrics_collector.sh');
+        const remoteScriptsDir: string = this.getRemoteScriptsPath();
+        const gpuCollectorContent: string = String.Format(
+            GPU_COLLECTOR_FORMAT, 
+            remoteScriptsDir, 
+            path.join(remoteScriptsDir, 'pid'), 
+        );
+        await fs.promises.writeFile(gpuMetricFilePath, gpuCollectorContent, { encoding: 'utf8' });
+        return Promise.resolve(localGpuMetricFolder);
```

Comment:
Since the function name is generateGpuScript(),  normally it should return nothing,  worse choice is to return the full path of the script,   returning a folder is weird.

Reply:
set full path in a function, and generateGpuScript() returns nothing.